aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,567 @@
+            return Err(Error::InvalidConfiguration(
+                "depth must be greater than zero",
+            ));
+        } else if concurrency > depth {
+            // We enforce this property otherwise it leads to strange intermittent behavior, i.e.
+            // rejecting requests due to the queue being full even though there's sufficient
+            // concurrency slots open.
+            return Err(Error::InvalidConfiguration(
+                "depth must be greater than or equal to concurrency",
+            ));
+        }
+
+        let stack: Stack = Default::default();
+        let notifier = Arc::new(Notify::new());
+        let (cancel_tx, cancel_rx) = mpsc::unbounded_channel();
+        let semaphore = Arc::new(Semaphore::new(concurrency));
+        let dispatcher = Self::dispatcher(
+            Arc::clone(&stack),
+            Arc::clone(&notifier),
+            Arc::clone(&semaphore),
+        );
+        let canceller = Self::canceller(Arc::clone(&stack), cancel_rx);
+        let id_generator = Default::default();
+        Ok(Self {
+            depth,
+            stack,
+            notifier,
+            _semaphore: semaphore,
+            dispatcher,
+            id_generator,
+            cancel_tx,
+            canceller,
+        })
+    }
+
+    fn dispatcher(
+        stack: Stack,
+        notifier: Arc<Notify>,
+        semaphore: Arc<Semaphore>,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            loop {
+                let semaphore = Arc::clone(&semaphore);
+
+                // Unwrap safety: the semaphore is never dropped.
+                let permit = semaphore.acquire_owned().await.unwrap();
+                let dispatcher = {
+                    // Unwrap safety: mutex poisoning
+                    let mut stack = stack.lock().unwrap();
+                    stack.pop()
+                };
+                match dispatcher {
+                    Some(mut dispatcher) => {
+                        dispatcher.send(permit);
+                    }
+                    None => {
+                        // The queue is empty. Sleep here until we're notified that there's another
+                        // ticket available to work on.
+                        notifier.notified().await;
+                    }
+                }
+            }
+        })
+    }
+
+    fn canceller(stack: Stack, mut cancel_rx: mpsc::UnboundedReceiver<TicketId>) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            while let Some(cancellation) = cancel_rx.recv().await {
+                // Unwrap safety: mutex poisoning.
+                let mut stack = stack.lock().unwrap();
+
+                // For a Vec, this is worst case O(n) per cancelled request. We anticipate the
+                // queue depth will be small, so cache locality should keep this relatively
+                // performant.
+                stack.retain(|dispatcher| dispatcher.id != cancellation);
```

Comment:
If `depth` is large, this will be painful -- we are holding the stack lock while this occurs, so we won't be making any progress in the `dispatcher` or any `acquire` attempts.

A few options:
1) Using a linked list would allow `O(1)` for all relevant operations (enqueueing, dequeueing, removal from arbitrary location); however, amazingly, the stdlib `LinkedList` does not allow `O(1)` removal from arbitrary locations [outside of the nightly-only `Cursor` API]. Unless we want to roll our own linked list, I suppose this is out.

2) Rather than removing items immediately, we could mark `Dispatcher`s as cancelled, then sweep at opportune moments (i.e. when attempting to enqueue when the queue is full), effectively amortizing the cost of removal due to cancellation. When popping the stack, we would pop until we found an uncancelled element. A race-free way to determine if a `Dispatcher` is cancelled would be to check if the send channel is closed when we try to `send` to it, since dropping the `Receiver` automatically closes the `Sender`.

3) We could do the swap-remove trick to remove items in `O(1)` time, at the cost of only being _approximately_ LIFO. This might be the simplest solution. (By the "swap-remove trick", I mean swap the element to be removed with the first/last element, then remove it -- both operations are constant-time.)

Reply:
Good call out on BTreeMap.

> We could accept that overflow means some queued requests will be handled out-of-order

Considering practicality, if we overflow at 2^64 requests, then it'll take us approximately 584,942 years at 1 million QPS to overflow this counter. (Please double check my math :D).

That wouldn't be the longest I've seen a system stay up without receiving patches 😉, but I think we can accept the tradeoff.

Overflow happens on 32-bits in about an hour at 1M QPS, but I think we can just leave the ticket ID as a u64.